doc/contributing/erm-guidelines.md

```diff
@@ -0,0 +1,307 @@
+  // there was a pending exception already.
+  if (error instanceof SuppressedError) {
+    console.error('An error occurred during disposal masking pending error:',
+                  error.error, error.suppressed);
+  } else {
+    console.error('An error occurred:', error);
+  }
+}
+```
+
+## Guidelines for Introducing ERM into Existing APIs
+
+Introducing ERM capabilities into existing APIs can be tricky.
+
+The best way to understand the issues is to look at a real world example. PR
+[58516](https://github.com/nodejs/node/pull/58516) is a good case. This PR
+sought to introduce ERM capabilities into the `fs.mkdtemp` API such that a
+temporary directory could be created and automatically disposed of when the
+scope in which it was created exited. However, the existing implementation of
+the `fs.mkdtemp` API returns a string value that cannot be made disposable.
+There are also sync, callback, and promise-based variations of the existing
+API that further complicate the situation.
+
+In the initial proposal, the `fs.mkdtemp` API was changed to return an object
+that implements the `Symbol.dispose` method but only if a specific option is
+provided. This would mean that the return value of the API would become
+polymorphic, returning different types based on how it was called. This adds
+a lot of complexity to the API and makes it difficult to reason about the
+return value. It also makes it difficult to programmatically detect whether
+the version of the API being used supports ERM capabilities or not. That is,
+`fs.mkdtemp('...', { dispoable: true })` would act differently in older versions
+of Node.js than in newer versions with no way to detect this at runtime other
+than to inspect the return value.
+
+Some APIs that already return object that can be made disposable do not have
+this kind of issue. For example, the `setTimeout` API in Node.js returns an
+object that implements the `Symbol.dispose` method. This change was made without
+much fanfare because the return value of the API was already an object.
+
+So, some APIs can be made disposable easily without any issues while others
+require more thought and consideration. The following guidelines can help
+when introducing ERM capabilities into existing APIs:
+
+1. Avoid polymorphic return values: If an API already returns a value that
+   can be made disposable and it makes sense to make it disposable, do so. Do
+   not, however, make the return value polymorphic determined by an option
+   passed into the API.
+2. Introduce new API variants that are ERM capable: If an existing API
+   cannot be made disposable without changing the return type or making it
+   polymorphic, consider introducing a new API variant that is ERM capable.
+   For example, `fs.mkdtempDisposable` could be introduced to return a
+   disposable object while the existing `fs.mkdtemp` API continues to return
+   a string. Yes, it means more APIs to maintain but it avoids the complexity
```

Comment:
Could/should we also have a deprecation strategy/guideline in this case?

For example, `the introduction of disposable variants should whenever possible be accompanied by the deprecation of their non-disposable equivalent`?

In that way this could help not increasing the APIs surface that the user needs to be aware of and we need to maintain?

Reply:
Hmm.. good question. I really don't know.